File: product-service/src/common/midlleware/request-id.ts
import { v4 as uuid } from 'uuid';
import { Request, Response, NextFunction } from 'express';

export function RequestIdMiddleware(req: Request, res: Response, next: NextFunction) {
  const requestId = req.headers['x-request-id'] || uuid();
  req['requestId'] = requestId;
  res.setHeader('X-Request-ID', requestId);
  console.log(`[Middleware] Assigned requestId: ${requestId}`);
  next();
}

File: order-service/src/middleware/request-id.go
package middleware

import (
	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

// AddRequestID adds a unique request ID to each request and response header
func AddRequestID() fiber.Handler {
	return func(c *fiber.Ctx) error {
		requestID := c.Get("X-Request-ID")
		if requestID == "" {
			requestID = uuid.NewString()
		}

		c.Locals("request_id", requestID)
		c.Set("X-Request-ID", requestID)

		return c.Next()
	}
}